chore: add contributing file (#65)
microsoft/fast-tooling: CONTRIBUTING.md

# Contributing to FAST Tooling

## Getting started

### Machine setup

> To work with the FAST Tooling [monorepo](https://en.wikipedia.org/wiki/Monorepo) you'll need Git, Node.js, Yarn and Lerna setup on your machine.

Reviewer: ```suggestion
To work with the FAST Tooling [monorepo](https://en.wikipedia.org/wiki/Monorepo) you'll need Git, Node.js, and NPM setup on your machine.
```

FAST Tooling uses Git as its source control system. If you haven't already installed it, you can download it [here](https://git-scm.com/downloads) or if you prefer a GUI-based approach, try [GitHub Desktop](https://desktop.github.com/).

Once Git is installed, you'll also need Node.js, which FAST Tooling uses as its JavaScript runtime, enabling its build and test scripts. Node.js instructions and downloads for your preferred OS can be found [here](https://nodejs.org/en/).

> Because the FAST Tooling repository is structured as a monorepo, we'll need a couple of tools to manage that. The first is Yarn, which can be installed by executing the following command at the terminal:
> 
> ```shell
> npm install -g yarn
> ```
> 
> The second tool you'll need is Lerna, which can be installed with this command:
> 
> ```bash
> yarn global add lerna
> ```
> 

Reviewer: ```suggestion
```

:::important
The above steps are a one-time setup for your machine and do not need to be repeated after the initial configuration.
:::

### Cloning the repository

Now that your machine is setup, you can clone the FAST Tooling repository. Open a terminal and run this command:

```shell
git clone https://github.com/microsoft/fast-tooling.git
```
Cloning via SSH:

```shell
git clone git@github.com:microsoft/fast-tooling.git
```

### Installing and building

From within the `fast-tooling` folder where you've cloned the repo, install all package dependencies and build all workspaces (local dependencies) with this command:

```bash

> yarn

Reviewer: ```suggestion
npm install
```

```

After the initial install, you can re-build all workspaces in the future with:

```bash

> yarn prepare

Reviewer: ```suggestion
npm run prepare --workspaces
```

```

### Testing

To run all tests for all packages, use the following command:

```bash

> yarn test

Reviewer: ```suggestion
npm run test --workspaces
```

```

This command can also be run from within individual package folders to execute only tests from that package.

:::note
Packages are located within the `packages` folder of the repository. Each package has a `package.json` file with a `scripts` section that defines the commands available to you for common tasks such as build, test, lint, etc.
:::

### Submitting a pull request

> If you'd like to contribute by fixing a bug, implementing a feature, or even correcting typos in our documentation, you'll want to submit a pull request. Before submitting a pull request, be sure to [rebase](https://www.atlassian.com/git/tutorials/merging-vs-rebasing) your branch from master or use the *merge* button provided by GitHub.

Reviewer: ```suggestion
If you'd like to contribute by fixing a bug, implementing a feature, or even correcting typos in our documentation, you'll want to submit a pull request. Before submitting a pull request, be sure to [rebase](https://www.atlassian.com/git/tutorials/merging-vs-rebasing) your branch from main or use the *merge* button provided by GitHub.
```

When submitting your pull request please make the title clear and concise, provide a description of the change, and specify any issues that will be closed.

#### Change Files

> Any pull request which includes changes within the `packages/*` directory requires a corresponding change file. Before pushing your changes to create a pull request, be sure you have included the necessary change file(s). To generate a change file, run `yarn change` in the root of the repository. The generated file will be checked into the repo automatically for you as part of the process.

Reviewer: ```suggestion
Any pull request which includes changes within the `packages/*` directory requires a corresponding change file. Before pushing your changes to create a pull request, be sure you have included the necessary change file(s). To generate a change file, run `npm change` in the root of the repository. The generated file will be checked into the repo automatically for you as part of the process.
```

:::note

> When working across feature branches, you'll need to target the branch using the following command: `yarn change --branch origin/{branch-name}`.

Reviewer: ```suggestion
When working across feature branches, you'll need to target the branch using the following command: `npm change --branch origin/{branch-name}`.
```

:::

**Example: Generated change file:**
```json
{
  "type": "minor",

>   "comment": "add fancy new feature for components",

Reviewer: ```suggestion
  "comment": "Add fancy new feature for components.",
```

  "packageName": "@microsoft/fast-tooling",
  "email": "name@example.com",
  "dependentChangeType": "minor",
  "date": "2021-03-01T19:10:06.323Z"
}
```

> Running `yarn change` will walk you through a CLI process for generating change files. The process will walk you through selecting the type of change as well as ask you to provide a description of any changes. As a convenience, the utiliity looks to provide recent commit messages for use in the description. *For changes which do not affect the published package(s), please use "none" when selecting the change type*.

Reviewer: ```suggestion
Running `npm change` will walk you through a CLI process for generating change files. The process will walk you through selecting the type of change as well as ask you to provide a description of any changes. As a convenience, the utility looks to provide recent commit messages for use in the description. *For changes which do not affect the published package(s), please use "none" when selecting the change type*.
```
